test/integration/framework/perf_utils.go

```diff
@@ -89,8 +94,9 @@ func (p *IntegrationTestNodePreparer) PrepareNodes(nextNodeIndex int) error {
 	for i := 0; i < numNodes; i++ {
 		var err error
 		for retry := 0; retry < retries; retry++ {
-			_, err = p.client.CoreV1().Nodes().Create(context.TODO(), baseNode, metav1.CreateOptions{})
+			node, err := p.client.CoreV1().Nodes().Create(context.TODO(), baseNode, metav1.CreateOptions{})
 			if err == nil {
+				p.preparedNodes.Insert(node.Name)
```

Comment:
I'm not fully familiar with this test , but in Lin108 it lists all the nodes again ...

My question is, why the test intersect with other nodes and is not properly isolated? it sounds weird that an integration test create nodes in paralell with other tests, usually each tests creates its own apiserver with its own etcd shard, that guarantees its independence ... and everything is destroyed on shutdown

Reply:
Currently we may have multi IntegrationTestNodePreparer instances in one cluster, so functionally, they should only cleanup the nodes created by themself. I wouldn't say this is a good design, but maintaining only one manager sounds better.

This is based on my current understanding. I just visit these codes recently.

Reply:
> Currently we may have multi IntegrationTestNodePreparer instances in one cluster

Do we? It sounds like something that shouldn't be allowed. So we could produce a fatal if we create more than one.

Reply:
Yea, @kerthcet is correct. `IntegrationTestNodePreparer` is created every time `createNodesOp` is operated and it works.

https://github.com/kubernetes/kubernetes/blob/8287e21228e9b1c945d031581e2fc63920f46d56/test/integration/scheduler_perf/scheduler_perf_test.go#L668

`getNodePreparer` creates `IntegrationTestNodePreparer`
https://github.com/kubernetes/kubernetes/blob/8287e21228e9b1c945d031581e2fc63920f46d56/test/integration/scheduler_perf/scheduler_perf_test.go#L914

Reply:
That's why cleaning up only Nodes prepared by that `IntegrationTestNodePreparer` makes sense.

But, yeah, I think using only one `IntegrationTestNodePreparer` and cleaning up all Nodes in a `CleanupNodes` method would be another nice approach. And, actually, I prefer it to the current multiple-`IntegrationTestNodePreparer` implementation. (require a little big change though.)

Reply:
Thanks @sanposhiho Do we have any plan to refactor this as a singleton object or this is as designed?

Reply:
I got the answer. Thanks.

Reply:
oh, commented at the same time 😄

Reply:
Sorry, what is the conclusion?

Reply:
The conclusion is for the current, we may have multi `IntegrationTestNodePreparer` instances, so this PR walks on the right way, just cleaning up the created nodes by its own. 

In the future, we'll refactor the architecture to make `IntegrationTestNodePreparer` a singleton, then we'll cleanup all nodes in this test. But I think it still make sense for the nodes to cleanup are also the created ones.

Reply:
"we may" or "we do"?

I think the requirement shouldn't be that we call the preparer once. The requirement is that we shouldn't call it again, if we haven't cleaned the nodes first.

I prefer we move in that direction instead of merging this PR.

/hold

Reply:
we should not have test pollution if we can ...

Reply:
> "we may" or "we do"?
> 
It depends, if he/she adds multi `createNodesOp` in yaml file, we'll create multi instances.

Reply:
> I think the requirement shouldn't be that we call the preparer once. 

My first design in my mind is that we may instantiate one preparer, and register nodes to the preparer when we call `createNodesOp`, tear down when finished. So I don't care about how many times you call, but we'll have only one manager.

> The requirement is that we shouldn't call it again, if we haven't cleaned the nodes first.

If we can make sure that we can clean up the nodes created by each preparer instance finally,  it doesn't matter anymore. 

And we may also have the cases like different node types for different preparer instances. Then we should keep them both.

Reply:
> we should not have test pollution if we can ...

Absolutely, what test pollution do you mean, between different tests?

Reply:
I'm ok to close this PR for it actually doesn't make a lot of improvements.